openmc/data/resonance_covariance.py

```diff
@@ -0,0 +1,743 @@
+                records.append([energy[i], spin[i], gt[i], gn[i],
+                                gg[i], gf[i]])
+
+            corr = endf.get_intg_record(file_obj)
+            cov = np.diag(par_unc).dot(corr).dot(np.diag(par_unc))
+
+            # Create pandas DataFrame with resonacne data
+            columns = ['energy', 'J', 'totalWidth', 'neutronWidth',
+                       'captureWidth', 'fissionWidth']
+            parameters = pd.DataFrame.from_records(records, columns=columns)
+
+            # Determine mpar (number of parameters for each resonance in
+            # covariance matrix)
+            nparams, params = parameters.shape
+            covsize = cov.shape[0]
+            mpar = int(covsize/nparams)
+
+            # Add parameters from File 2
+            parameters = _add_file2_contributions(parameters,
+                                                  resonance.parameters)
+
+        elif lcomp == 0:
+            cov = np.zeros([4, 4])
+            records = []
+            cov_index = 0
+            for i in range(nls):
+                items, values = endf.get_list_record(file_obj)
+                num_res = items[5]
+                for j in range(num_res):
+                    one_res = values[18*j:18*(j+1)]
+                    res_values = one_res[:6]
+                    cov_values = one_res[6:]
+                    records.append(list(res_values))
+
+                    # Populate the coviariance matrix for this resonance
+                    # There are no covariances between resonances in lcomp=0
+                    cov[cov_index, cov_index] = cov_values[0]
+                    cov[cov_index+1, cov_index+1 : cov_index+2] = cov_values[1:2]
+                    cov[cov_index+1, cov_index+3] = cov_values[4]
+                    cov[cov_index+2, cov_index+2] = cov_values[3]
+                    cov[cov_index+2, cov_index+3] = cov_values[5]
+                    cov[cov_index+3, cov_index+3] = cov_values[6]
+
+                    cov_index += 4
+                    if j < num_res-1:  # Pad matrix for additional values
+                        cov = np.pad(cov, ((0, 4), (0, 4)), 'constant',
+                                     constant_values=0)
+
+            # Create pandas DataFrame with resonance data, currently
+            # redundant with data.IncidentNeutron.resonance
+            columns = ['energy', 'J', 'totalWidth', 'neutronWidth',
+                       'captureWidth', 'fissionWidth']
+            parameters = pd.DataFrame.from_records(records, columns=columns)
+
+            # Determine mpar (number of parameters for each resonance in
+            # covariance matrix)
+            nparams, params = parameters.shape
+            covsize = cov.shape[0]
+            mpar = int(covsize/nparams)
+
+            # Add parameters from File 2
+            parameters = _add_file2_contributions(parameters,
+                                                  resonance.parameters)
```

Comment:
It looks like the last few lines here (starting with `columns = ...`) are the same for each branch of the if -- I think those could be moved out of the if altogether (so that they're not repeated). The only thing that's different is in the lcomp == 1 case, `mpar` is not set, but that looks like a bug since it is used in the call below.